network/src/protocols/rpc/mod.rs

```diff
@@ -383,6 +384,13 @@ async fn handle_inbound_substream<TSubstream>(
 ) where
     TSubstream: AsyncRead + AsyncWrite + Debug + Send + Unpin,
 {
+    debug_checked_precondition!({
+        match notif {
+            PeerManagerNotification::NewInboundSubstream(_, _) => true,
+            _ => false,
+        }
+    });
+
```

Comment:
won't the `unreachable!()` branch below still panic in production?

Reply:
yup, I'm experimenting with this PR :D I think the final solution is going to change `unreachable` to `debug_assert` and avoid MIRAI.

Reply:
The unreachable!() will only panic in production if it is actually reachable. The precondition precludes that possibility. Of course, the precondition needs to be enforced. If you only enforce it at runtime then you just move the runtime failure from the unreachable!() annotation to the precondition, which does not solve the problem that David is setting out to solve in the first place.

So there are two options here:
1) Replace the unreachable!() call with something that does not fail during production, even it is reached. One way to that is debug_assert!. Downside: while the node will not crash we may be missing something important when this code is reached. In the production systems I'm familiar with, there would be output to a log and the log would be monitored on a regular basis.

2) Enforce the precondition at compile time. This way, we will be quite sure that no call site would ever call this function with arguments that could cause the unreachable!() statement to actually be reached. This is the true meaning of unreachable!, namely that the programmer certifies that it will never be reached. The downside of this is programmers make mistakes, so we are not exactly quite sure. This is where MIRAI helps: If you run MIRAI as part of CI, then you have a mechanical proof that the precondition is never violated.

Coming back to the proposed change to use debug_assert. The first version of this PR used debug_checked_verify!(false, "..."). That is equivalent to debug_assert at runtime, but when MIRAI runs, it complains that an assertion that is statically known to be false is likely a programmer mistake. After all, if you mean so say "unreachable", well there is a macro for just that.

In the current state of things where MIRAI is not run in CI and where MIRAI itself is not completely trusted, the debug_assert is the best compromise: It does not fail in production, even if MIRAI is wrong and it also allows MIRAI automatically infer and enforce (at compile time) the precondition. Moreover it is less verbose than other options.

I would, however, strongly recommend that the debug_assert is complemented by a log event.

Reply:
Indeed. For context, the rpc actor is (currently) never registered to receive New/LostPeer events, which is why I put the `unreachable!()` there. There is actually no problem with just ignoring the events, but I would want CI to fail if that started happening to signal some change in the model. Hence I agree that a `debug_assert!()` is likely the best compromise.

Reply:
can't we have MIRAI check that `debug_assert!(false)` is never reached?

Reply:
If you run MIRAI over this code it will report a warning if it concludes that the debug_assert is reachable.

Reply:
OK so I understand this as `debug_assert!(false)` is the right way to go, and in the future MIRAI might even be able to tell us that this might get hit.